src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,392 @@
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGR : GL_RGB;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    case QImage::Format_RGBA8888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGRA : GL_RGBA;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    default:
+        m_isValid = false;
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pThread = std::make_unique<QThread>();
+    m_pThread->setObjectName("ControllerScreenRenderer");
+
+#if QT_VERSION >= QT_VERSION_CHECK(6, 7, 0)
+    [[maybe_unused]] bool successful = moveToThread(m_pThread.get());
+    DEBUG_ASSERT(successful);
+#else
+    moveToThread(m_pThread.get());
+#endif
+
+    // these at first sight weird-looking connections facilitate thread-safe communication.
+    connect(this,
+            &ControllerRenderingEngine::engineSetupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::sendFrameDataRequested,
+            this,
+            &ControllerRenderingEngine::send);
+    connect(m_pThread.get(),
+            &QThread::finished,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_pThread->start(QThread::NormalPriority);
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    DEBUG_ASSERT(QThread::currentThread() != thread());
+    m_pThread->wait();
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.critical() << "The ControllerEngine is being deleted but hasn't been "
+                              "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!thread()->isFinished() && !thread()->isInterruptionRequested()) {
+        kLogger.critical() << "Render thread has or is about to terminate. Cannot "
+                              "start this render anymore.";
+        return;
+    }
+    QCoreApplication::postEvent(this, new QEvent(QEvent::UpdateRequest));
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pThread && m_pThread->isRunning();
+}
+
+void ControllerRenderingEngine::requestEngineSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
+    VERIFY_OR_DEBUG_ASSERT(qmlEngine) {
+        kLogger.critical() << "No QML engine was passed!";
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(QThread::currentThread() != thread()) {
+        kLogger.warning() << "Unable to setup OpenGL rendering context from the same "
+                             "thread as the render object";
+        return;
+    }
+    emit engineSetupRequested(qmlEngine);
+
+    const auto lock = lockMutex(&m_mutex);
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+    if (m_isValid) {
+        m_renderControl->prepareThread(m_pThread.get());
+    }
+}
```

Comment:
this is called from another thread, right? then setting `m_isValid` is already a potential data race. 

Also the `m_mutex` and waitcondition is only really there to ensure `requestEngineSetup` effectively runs synchronously with `setup`, right? So the mutex could be avoided if the connection that connected `engineSetupRequested` was a Qt::BlockingQueuedConnection. Moreover, you don't even need the connection and the signal if you use `QMetaObject::invokeMethod` drastically simplifying the code. Am I missing some flaw with that approach?

Reply:
That's much more elegant and appears to work fine. Thanks!